core/src/main/java/org/mule/TransformationService.java

```diff
@@ -0,0 +1,486 @@
+                }
+            }
+        }
+
+        if (previous.getOutboundAttachmentNames().size() > 0)
+        {
+            for (String name : previous.getOutboundAttachmentNames())
+            {
+                try
+                {
+                    target.addOutboundAttachment(name, previous.getOutboundAttachment(name));
+                }
+                catch (Exception e)
+                {
+                    throw new MuleRuntimeException(CoreMessages.failedToReadAttachment(name), e);
+                }
+            }
+        }
+    }
+
+    /**
+     * Will attempt to obtain the payload of this message with the desired Class type. This will
+     * try and resolve a transformer that can do this transformation. If a transformer cannot be
+     * found an exception is thrown. Any transformers added to the registry will be checked for
+     * compatability.
+     *
+     * @param resultType the desired return type
+     * @param encoding   the encoding to use if required
+     * @return The converted payload of this message. Note that this method will not alter the
+     *         payload of this message <b>unless</b> the payload is an {@link InputStream} in which
+     *         case the stream will be read and the payload will become the fully read stream.
+     * @throws TransformerException if a transformer cannot be found or there is an error during
+     *                              transformation of the payload.
+     * @since 3.0
+     */
+    @SuppressWarnings("unchecked")
+    private  <T> T getPayload(MuleMessage message, DataType<T> resultType, String encoding) throws TransformerException
+    {
+        // Handle null by ignoring the request
+        if (resultType == null)
+        {
+            throw new IllegalArgumentException(CoreMessages.objectIsNull("resultType").getMessage());
+        }
+
+        DataType source = DataTypeFactory.createFromObject(message);
+
+        // If no conversion is necessary, just return the payload as-is
+        if (resultType.isCompatibleWith(source))
+        {
+            return (T) message.getPayload();
+        }
+
+        // The transformer to execute on this message
+        Transformer transformer = muleContext.getRegistry().lookupTransformer(source, resultType);
+        if (transformer == null)
+        {
+            throw new TransformerException(CoreMessages.noTransformerFoundForMessage(source, resultType));
+        }
+
+        // Pass in the message itself
+        Object result = transformer.transform(message, encoding);
+
+        // Unless we disallow Object.class as a valid return type we need to do this extra check
+        if (!resultType.getType().isAssignableFrom(result.getClass()))
+        {
+            throw new TransformerException(CoreMessages.transformOnObjectNotOfSpecifiedType(resultType, result));
+        }
+
+        // If the payload is a stream and we've consumed it, then we should set the payload on the
```

Comment:
Seems to me that if the payload is consumable, then the new value should be set on the message before throwing the previous exception